ccl/sqlproxyccl: TestDirectoryConnect failed

ccl/sqlproxyccl.TestDirectoryConnect [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3193326&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3193326&tab=artifacts#/) on master @ [1487f4fc619d5d62cbb29962ae2b6b02dba432ee](https://github.com/cockroachdb/cockroach/commits/1487f4fc619d5d62cbb29962ae2b6b02dba432ee):


```
I210716 22:21:01.968019 8561 2@rpc/nodedialer/nodedialer.go:161  [intExec=read orphaned leases,sql] 521  unable to connect to n1: failed to connect to n1 at 127.250.40.147:34769: context canceled
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522  unable to read orphaned leases: collecting referenced descriptors: failed to send RPC: sending to all replicas failed; last error: failed to connect to n1 at 127.250.40.147:34769: context canceled
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +(1) keys: [sql.schema.validation_errors.write.read_referenced_descriptors]
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +Wraps: (2) keys: [sql.schema.validation_errors.write.read_referenced_descriptors]
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +Wraps: (3) attached stack trace
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  -- stack trace:
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog.(*validationErrorAccumulator).decorate
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/validate.go:289
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog.(*validationErrorAccumulator).reportDescGetterError
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/validate.go:282
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog.Validate
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/validate.go:57
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*kvDescriptors).validateUncommittedDescriptors
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/kv_descriptors.go:358
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).ValidateUncommittedDescriptors
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:257
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).commitSQLTransactionInternal
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:807
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).commitSQLTransaction
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:790
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).handleAutoCommit
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1530
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState.func7
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:498
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:728
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:126
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1618
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1620
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1542
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:204
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | runtime.goexit
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | 	/usr/local/go/src/runtime/asm_amd64.s:1371
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +Wraps: (4) collecting referenced descriptors
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +Wraps: (5) failed to send RPC: sending to all replicas failed; last error: failed to connect to n1 at 127.250.40.147:34769: context canceled
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  |
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | (opaque error leaf)
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +  | type name: github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/kvcoord.sendError
W210716 22:21:01.968340 8557 sql/catalog/lease/lease.go:1095  [-] 522 +Error types: (1) *telemetrykeys.withTelemetry (2) *telemetrykeys.withTelemetry (3) *withstack.withStack (4) *errutil.withPrefix (5) *errbase.opaqueLeaf
W210716 22:21:01.995151 8603 kv/txn.go:667  [-] 523  failure aborting transaction: node unavailable; try another peer; abort caused by: could not query jobs table: cancel/pause-requested: context canceled
E210716 22:21:01.995305 8603 jobs/registry.go:623  [-] 524  failed to serve pause and cancel requests: could not query jobs table: cancel/pause-requested: context canceled
W210716 22:21:01.995371 8603 jobs/registry.go:664  [-] 525  canceling all adopted jobs due to stopper quiescing
W210716 22:21:01.999210 6928 sql/sqlliveness/slinstance/slinstance.go:183  [n1] 526  exiting heartbeat loop
W210716 22:21:01.999388 6929 jobs/registry.go:664  [-] 527  canceling all adopted jobs due to stopper quiescing
    --- FAIL: TestDirectoryConnect/drain_connection (0.00s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestDirectoryConnect PKG=./pkg/ccl/sqlproxyccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server andy-kimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDirectoryConnect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDirectoryConnect [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3194866&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3194866&tab=artifacts#/) on master @ [efb4fd2b406cd47b505b4f1a1d5700fc79841469](https://github.com/cockroachdb/cockroach/commits/efb4fd2b406cd47b505b4f1a1d5700fc79841469):


```
W210719 05:07:43.129889 6663 kv/range_lookup.go:243  [n1,rangefeed=table-stats-cache,range-lookup=/Meta2/Table/20/NULL] 421  range lookup of key /Meta2/Table/20/NULL found only non-matching ranges []; retrying
W210719 05:07:43.135921 6662 kv/range_lookup.go:243  [n1,rangefeed=table-stats-cache,range-lookup=/Table/20] 422  range lookup of key /Table/20 found only non-matching ranges []; retrying
I210719 05:07:43.138489 6544 kv/kvserver/replica_rangefeed.go:619  [n1,rangefeed=table-stats-cache,s1,r16/1:/Table/2{0-1}] 423  RangeFeed closed timestamp is empty
I210719 05:07:43.143178 6687 kv/kvserver/queue.go:583  [n1,s1,r27/1:/{NamespaceTab…-Table/32}] 424  rate limited in MaybeAdd (merge): throttled on async limiting semaphore
I210719 05:07:43.147928 6582 1@util/log/event_log.go:32  [n1] 425 ={"Timestamp":1626671263147924938,"EventType":"node_join","NodeID":1,"StartedAt":1626671263101237737,"LastUp":1626671263101237737}
I210719 05:07:43.148111 6582 sql/sqlliveness/slinstance/slinstance.go:252  [n1] 426  starting SQL liveness instance
I210719 05:07:43.149628 6856 sql/temporary_schema.go:492  [n1] 427  running temporary object cleanup background job
I210719 05:07:43.150590 6856 sql/temporary_schema.go:529  [n1] 428  found 0 temporary schemas
I210719 05:07:43.150647 6856 sql/temporary_schema.go:532  [n1] 429  early exiting temporary schema cleaner as no temporary schemas were found
I210719 05:07:43.150672 6856 sql/temporary_schema.go:533  [n1] 430  completed temporary object cleanup job
I210719 05:07:43.150691 6856 sql/temporary_schema.go:612  [n1] 431  temporary object cleaner next scheduled to run at 2021-07-19 05:37:43.149577125 +0000 UTC
I210719 05:07:43.159319 6978 util/log/event_log.go:32  [n1,intExec=optInToDiagnosticsStatReporting] 432 ={"Timestamp":1626671263152241564,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"diagnostics.reporting.enabled\" = true","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-optInToDiagnosticsStatReporting","SettingName":"diagnostics.reporting.enabled","Value":"true"}
I210719 05:07:43.164848 7084 migration/migrationmanager/manager.go:102  [n1,intExec=set-setting,migration-mgr] 433  no need to migrate, cluster already at newest version
I210719 05:07:43.168361 7084 util/log/event_log.go:32  [n1,intExec=set-setting] 434 ={"Timestamp":1626671263162861136,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING version = $1","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-set-setting","PlaceholderValues":["'21.1-116'"],"SettingName":"version","Value":"21.1-116"}
I210719 05:07:43.179793 6964 sql/sqlliveness/slstorage/slstorage.go:386  [n1] 435  inserted sqlliveness session c838ac2a9872412fb9164b4320140c58
I210719 05:07:43.191900 6964 sql/sqlliveness/slinstance/slinstance.go:144  [n1] 436  created new SQL liveness session c838ac2a9872412fb9164b4320140c58
I210719 05:07:43.212297 7259 util/log/event_log.go:32  [n1,intExec=initializeClusterSecret] 437 ={"Timestamp":1626671263206318300,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"cluster.secret\" = gen_random_uuid()::STRING","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-initializeClusterSecret","SettingName":"cluster.secret","Value":"8d1830de-a71d-41f7-af09-ef6648a92d10"}
I210719 05:07:43.224092 7408 5@util/log/event_log.go:32  [n1,intExec=create-default-DB] 438 ={"Timestamp":1626671263219155652,"EventType":"create_database","Statement":"CREATE DATABASE IF NOT EXISTS defaultdb","Tag":"CREATE DATABASE","User":"root","DescriptorID":50,"ApplicationName":"$ internal-create-default-DB","DatabaseName":"defaultdb"}
I210719 05:07:43.237228 7556 5@util/log/event_log.go:32  [n1,intExec=create-default-DB] 439 ={"Timestamp":1626671263224762758,"EventType":"create_database","Statement":"CREATE DATABASE IF NOT EXISTS postgres","Tag":"CREATE DATABASE","User":"root","DescriptorID":51,"ApplicationName":"$ internal-create-default-DB","DatabaseName":"postgres"}
I210719 05:07:43.286415 6582 server/server_sql.go:931  [n1] 440  done ensuring all necessary startup migrations have run
I210719 05:07:43.286709 7745 jobs/job_scheduler.go:360  [n1] 441  waiting 4m0s before scheduled jobs daemon start
I210719 05:07:43.290200 7763 server/auto_upgrade.go:55  [n1] 442  no need to upgrade, cluster already at the newest version
I210719 05:07:43.349947 6687 3@vendor/github.com/cockroachdb/pebble/db.go:1530  [n1,pebble,s1] 443  [JOB 3] WAL created 000004
I210719 05:07:43.353017 6582 1@server/server.go:2262  [n1] 444  serving sql connections
I210719 05:07:43.359733 7727 kv/kvserver/replica_command.go:396  [n1,s1,r39/1:/{Table/43-Max}] 445  initiating a split of this range at key /Tenant/28 [r40] (manual)
I210719 05:07:44.130551 7464 kv/kvserver/replica_consistency.go:255  [n1,consistencyChecker,s1,r4/1:/System{/tsd-tse}] 446  triggering stats recomputation to resolve delta of {ContainsEstimates:2028 LastUpdateNanos:1626671263334086049 IntentAge:0 GCBytesAge:0 LiveBytes:-68627 LiveCount:-1014 KeyBytes:-52403 KeyCount:-1014 ValBytes:-16224 ValCount:-1014 IntentBytes:0 IntentCount:0 SeparatedIntentCount:0 SysBytes:0 SysCount:0 AbortSpanBytes:0}
I210719 05:07:44.372132 7927 ccl/sqlproxyccl/tenant/directory.go:336  [-] 447  established watch on pods
--- FAIL: TestDirectoryConnect (6.06s)
=== RUN   TestDirectoryConnect/drain_connection
I210719 05:07:45.997684 8512 sql/catalog/lease/lease.go:1113  [-] 514  released orphaned lease: {id:33 version:1 expiration:{Time:{wall:203018000 ext:63762268364 loc:<nil>}}}
I210719 05:07:45.999265 8464 ccl/sqlproxyccl/proxy_handler.go:385  [client=127.0.0.1:44612,cluster=tenant-cluster,tenant=28] 515  new connection
I210719 05:07:45.999292 8508 sql/catalog/lease/lease.go:1113  [-] 516  released orphaned lease: {id:4 version:1 expiration:{Time:{wall:76076000 ext:63762268365 loc:<nil>}}}
I210719 05:07:46.000454 8511 sql/catalog/lease/lease.go:1113  [-] 517  released orphaned lease: {id:23 version:1 expiration:{Time:{wall:177924000 ext:63762268352 loc:<nil>}}}
I210719 05:07:46.001067 8510 sql/catalog/lease/lease.go:1113  [-] 518  released orphaned lease: {id:15 version:1 expiration:{Time:{wall:59909000 ext:63762268378 loc:<nil>}}}
I210719 05:07:46.001197 8509 sql/catalog/lease/lease.go:1113  [-] 519  released orphaned lease: {id:12 version:1 expiration:{Time:{wall:325623000 ext:63762268368 loc:<nil>}}}
I210719 05:07:46.718547 7927 ccl/sqlproxyccl/tenant/directory.go:336  [-] 520  established watch on pods
I210719 05:07:47.006625 7927 ccl/sqlproxyccl/tenant/directory.go:417  [-] 521  deleted IP address 127.0.0.1:33049 for tenant 28
I210719 05:07:47.217644 8464 ccl/sqlproxyccl/proxy_handler.go:388  [client=127.0.0.1:44612,cluster=tenant-cluster,tenant=28] 522  closing after 1.22s
I210719 05:07:47.217943 8464 ccl/sqlproxyccl/server.go:176  [client=127.0.0.1:44612] 523  connection error: codeIdleDisconnect: idle connection closed
I210719 05:07:48.001555 8610 ccl/sqlproxyccl/server.go:176  [client=127.0.0.1:44730] 524  connection error: codeUnexpectedStartupMessage: unsupported post-TLS startup message: *pgproto3.CancelRequest
    proxy_handler_test.go:606: 
        	Error Trace:	proxy_handler_test.go:925
        	Error:      	Expected nil, but got: &sqlproxyccl.codeError{code:8, err:(*withstack.withStack)(0xc002a35098)}
        	Test:       	TestDirectoryConnect/drain_connection
I210719 05:07:48.002422 7927 ccl/sqlproxyccl/tenant/directory.go:336  [-] 525  established watch on pods
W210719 05:07:48.007918 8482 jobs/registry.go:664  [-] 526  canceling all adopted jobs due to stopper quiescing
W210719 05:07:48.008160 8427 ccl/kvccl/kvtenantccl/connector.go:171  [sql,tenant-connector] 527  error consuming GossipSubscription RPC: rpc error: code = Canceled desc = context canceled
W210719 05:07:48.015311 6965 jobs/registry.go:664  [-] 528  canceling all adopted jobs due to stopper quiescing
W210719 05:07:48.015688 6964 sql/sqlliveness/slinstance/slinstance.go:183  [n1] 529  exiting heartbeat loop
    --- FAIL: TestDirectoryConnect/drain_connection (2.01s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestDirectoryConnect PKG=./pkg/ccl/sqlproxyccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server andy-kimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDirectoryConnect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl.TestDirectoryConnect [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3209759&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3209759&tab=artifacts#/) on master @ [f0e2aa6abbbbf3318ea20e7dbcbe40819a809b83](https://github.com/cockroachdb/cockroach/commits/f0e2aa6abbbbf3318ea20e7dbcbe40819a809b83):


```
I210722 03:44:28.022318 6564 1@server/server.go:1725  [n1] 415  starting grpc server at 127.17.145.3:41887
I210722 03:44:28.022345 6564 1@server/server.go:1726  [n1] 416  advertising CockroachDB node at 127.17.145.3:41887
W210722 03:44:28.059751 6660 kv/range_lookup.go:243  [n1,rangefeed=table-stats-cache,range-lookup=/Meta2/Table/20/NULL] 417  range lookup of key /Meta2/Table/20/NULL found only non-matching ranges []; retrying
W210722 03:44:28.062032 6659 kv/range_lookup.go:243  [n1,rangefeed=table-stats-cache,range-lookup=/Table/20] 418  range lookup of key /Table/20 found only non-matching ranges []; retrying
I210722 03:44:28.064163 6569 kv/kvserver/replica_rangefeed.go:619  [n1,rangefeed=table-stats-cache,s1,r16/1:/Table/2{0-1}] 419  RangeFeed closed timestamp is empty
I210722 03:44:28.099319 6564 1@util/log/event_log.go:32  [n1] 420 ={"Timestamp":1626925468099316622,"EventType":"node_join","NodeID":1,"StartedAt":1626925468017905817,"LastUp":1626925468017905817}
I210722 03:44:28.099498 6564 sql/sqlliveness/slinstance/slinstance.go:252  [n1] 421  starting SQL liveness instance
I210722 03:44:28.099985 6556 sql/temporary_schema.go:492  [n1] 422  running temporary object cleanup background job
I210722 03:44:28.103311 6556 sql/temporary_schema.go:529  [n1] 423  found 0 temporary schemas
I210722 03:44:28.103372 6556 sql/temporary_schema.go:532  [n1] 424  early exiting temporary schema cleaner as no temporary schemas were found
I210722 03:44:28.103395 6556 sql/temporary_schema.go:533  [n1] 425  completed temporary object cleanup job
I210722 03:44:28.103414 6556 sql/temporary_schema.go:612  [n1] 426  temporary object cleaner next scheduled to run at 2021-07-22 04:14:28.099968134 +0000 UTC
I210722 03:44:28.103919 6888 sql/sqlliveness/slstorage/slstorage.go:386  [n1] 427  inserted sqlliveness session aa8c408d287e47e0b9183559cc06226d
I210722 03:44:28.104002 6888 sql/sqlliveness/slinstance/slinstance.go:144  [n1] 428  created new SQL liveness session aa8c408d287e47e0b9183559cc06226d
I210722 03:44:28.127799 6927 util/log/event_log.go:32  [n1,intExec=optInToDiagnosticsStatReporting] 429 ={"Timestamp":1626925468104711077,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"diagnostics.reporting.enabled\" = true","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-optInToDiagnosticsStatReporting","SettingName":"diagnostics.reporting.enabled","Value":"true"}
I210722 03:44:28.145312 4379 migration/migrationmanager/manager.go:102  [n1,intExec=set-setting,migration-mgr] 430  no need to migrate, cluster already at newest version
I210722 03:44:28.173083 4379 util/log/event_log.go:32  [n1,intExec=set-setting] 431 ={"Timestamp":1626925468133189059,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING version = $1","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-set-setting","PlaceholderValues":["'21.1-116'"],"SettingName":"version","Value":"21.1-116"}
I210722 03:44:28.228295 7431 util/log/event_log.go:32  [n1,intExec=initializeClusterSecret] 432 ={"Timestamp":1626925468206010213,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"cluster.secret\" = gen_random_uuid()::STRING","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-initializeClusterSecret","SettingName":"cluster.secret","Value":"c3b0c9c1-9b02-49d0-9937-b8e9d2622f46"}
I210722 03:44:28.234724 7523 5@util/log/event_log.go:32  [n1,intExec=create-default-DB] 433 ={"Timestamp":1626925468231178335,"EventType":"create_database","Statement":"CREATE DATABASE IF NOT EXISTS defaultdb","Tag":"CREATE DATABASE","User":"root","DescriptorID":50,"ApplicationName":"$ internal-create-default-DB","DatabaseName":"defaultdb"}
I210722 03:44:28.256714 7437 5@util/log/event_log.go:32  [n1,intExec=create-default-DB] 434 ={"Timestamp":1626925468235461072,"EventType":"create_database","Statement":"CREATE DATABASE IF NOT EXISTS postgres","Tag":"CREATE DATABASE","User":"root","DescriptorID":51,"ApplicationName":"$ internal-create-default-DB","DatabaseName":"postgres"}
I210722 03:44:28.288273 6564 server/server_sql.go:931  [n1] 435  done ensuring all necessary startup migrations have run
I210722 03:44:28.289280 7723 jobs/job_scheduler.go:360  [n1] 436  waiting 2m0s before scheduled jobs daemon start
I210722 03:44:28.294829 7725 server/auto_upgrade.go:55  [n1] 437  no need to upgrade, cluster already at the newest version
I210722 03:44:28.318962 6708 3@vendor/github.com/cockroachdb/pebble/db.go:1530  [n1,pebble,s1] 438  [JOB 3] WAL created 000004
I210722 03:44:28.329104 6564 1@server/server.go:2255  [n1] 439  serving sql connections
I210722 03:44:28.337754 7734 kv/kvserver/replica_command.go:396  [n1,s1,r39/1:/{Table/43-Max}] 440  initiating a split of this range at key /Tenant/28 [r40] (manual)
I210722 03:44:29.378270 7922 ccl/sqlproxyccl/tenant/directory.go:336  [-] 441  established watch on pods
--- FAIL: TestDirectoryConnect (11.19s)
=== RUN   TestDirectoryConnect/drain_connection
I210722 03:44:36.057722 8534 ccl/sqlproxyccl/proxy_handler.go:385  [client=127.0.0.1:59030,cluster=tenant-cluster,tenant=28] 516  new connection
I210722 03:44:36.072759 8588 sql/catalog/lease/lease.go:1125  [-] 517  released orphaned lease: {id:33 version:1 expiration:{Time:{wall:248557000 ext:63762522563 loc:<nil>}}}
I210722 03:44:36.079184 8585 sql/catalog/lease/lease.go:1125  [-] 518  released orphaned lease: {id:12 version:1 expiration:{Time:{wall:571179000 ext:63762522564 loc:<nil>}}}
I210722 03:44:36.084084 8587 sql/catalog/lease/lease.go:1125  [-] 519  released orphaned lease: {id:23 version:1 expiration:{Time:{wall:87427000 ext:63762522576 loc:<nil>}}}
I210722 03:44:36.088816 8584 sql/catalog/lease/lease.go:1125  [-] 520  released orphaned lease: {id:4 version:1 expiration:{Time:{wall:850589000 ext:63762522565 loc:<nil>}}}
I210722 03:44:36.092066 8586 sql/catalog/lease/lease.go:1125  [-] 521  released orphaned lease: {id:15 version:1 expiration:{Time:{wall:485947000 ext:63762522582 loc:<nil>}}}
I210722 03:44:37.081632 7922 ccl/sqlproxyccl/tenant/directory.go:417  [-] 522  deleted IP address 127.0.0.1:38921 for tenant 28
I210722 03:44:37.319400 8534 ccl/sqlproxyccl/proxy_handler.go:388  [client=127.0.0.1:59030,cluster=tenant-cluster,tenant=28] 523  closing after 1.26s
I210722 03:44:37.319672 8534 ccl/sqlproxyccl/server.go:176  [client=127.0.0.1:59030] 524  connection error: codeIdleDisconnect: idle connection closed
I210722 03:44:38.031258 6818 2@util/log/event_log.go:32  [n1] 525 ={"Timestamp":1626925478031253981,"EventType":"runtime_stats","MemRSSBytes":159932416,"GoroutineCount":377,"MemStackSysBytes":4128768,"GoAllocBytes":45418000,"GoTotalBytes":73352200,"GoStatsStaleness":0.00139179,"HeapFragmentBytes":7436784,"HeapReservedBytes":5210112,"HeapReleasedBytes":4915200,"CGoAllocBytes":2879968,"CGoTotalBytes":8093696,"CGoCallRate":0.0000015163571,"CPUUserPercent":0.00000036510585,"CPUSysPercent":0.000000062694944,"GCPausePercent":0.00000000323898,"GCRunCount":49,"NetHostRecvBytes":22958681,"NetHostSendBytes":22958681}
I210722 03:44:38.080908 8683 ccl/sqlproxyccl/server.go:176  [client=127.0.0.1:59220] 526  connection error: codeUnexpectedStartupMessage: unsupported post-TLS startup message: *pgproto3.CancelRequest
    proxy_handler_test.go:606: 
        	Error Trace:	proxy_handler_test.go:930
        	Error:      	Expected nil, but got: &sqlproxyccl.codeError{code:8, err:(*withstack.withStack)(0xc00229a138)}
        	Test:       	TestDirectoryConnect/drain_connection
I210722 03:44:38.081775 7922 ccl/sqlproxyccl/tenant/directory.go:336  [-] 527  established watch on pods
W210722 03:44:38.087229 8565 jobs/registry.go:664  [-] 528  canceling all adopted jobs due to stopper quiescing
W210722 03:44:38.087487 8497 ccl/kvccl/kvtenantccl/connector.go:171  [sql,tenant-connector] 529  error consuming GossipSubscription RPC: rpc error: code = Canceled desc = grpc: the client connection is closing
W210722 03:44:38.098195 6889 jobs/registry.go:664  [-] 530  canceling all adopted jobs due to stopper quiescing
W210722 03:44:38.098792 6888 sql/sqlliveness/slinstance/slinstance.go:183  [n1] 531  exiting heartbeat loop
    --- FAIL: TestDirectoryConnect/drain_connection (2.02s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestDirectoryConnect PKG=./pkg/ccl/sqlproxyccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-experience @cockroachdb/server andy-kimball
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDirectoryConnect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/sqlproxyccl/BUILD.bazel
+++ b/pkg/ccl/sqlproxyccl/BUILD.bazel
@@ -69,7 +69,6 @@ go_test(
         "//pkg/sql/pgwire",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
-        "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/testcluster",
         "//pkg/util/leaktest",

--- a/pkg/ccl/sqlproxyccl/frontend_admitter.go
+++ b/pkg/ccl/sqlproxyccl/frontend_admitter.go
@@ -28,21 +28,25 @@ var FrontendAdmit = func(
 	// hence it's important to close `conn` rather than `proxyConn` since closing
 	// the latter will not call `Close` method of `tls.Conn`.
 	var sniServerName string
-	// If we have an incoming TLS Config, require that the client initiates
-	// with a TLS connection.
+
+	// Read first message from client.
+	m, err := pgproto3.NewBackend(pgproto3.NewChunkReader(conn), conn).ReceiveStartupMessage()
+	if err != nil {
+		return conn, nil, newErrorf(codeClientReadFailed, "while receiving startup message")
+	}
+
+	// CancelRequest is unencrypted and unauthenticated, regardless of whether
+	// the server requires TLS connections. For now, ignore the request to cancel,
+	// and send back a nil StartupMessage, which will cause the proxy to just
+	// close the connection in response.
+	if _, ok := m.(*pgproto3.CancelRequest); ok {
+		return conn, nil, nil
+	}
+
+	// If we have an incoming TLS Config, require that the client initiates with
+	// an SSLRequest message.
 	if incomingTLSConfig != nil {
-		m, err := pgproto3.NewBackend(pgproto3.NewChunkReader(conn), conn).ReceiveStartupMessage()
-		if err != nil {
-			return conn, nil, newErrorf(codeClientReadFailed, "while receiving startup message")
-		}
-		switch m.(type) {
-		case *pgproto3.SSLRequest:
-		case *pgproto3.CancelRequest:
-			// Ignore CancelRequest explicitly. We don't need to do this but it
-			// makes testing easier by avoiding a call to SendErrToClient on this
-			// path (which would confuse assertCtx).
-			return conn, nil, nil
-		default:
+		if _, ok := m.(*pgproto3.SSLRequest); !ok {
 			code := codeUnexpectedInsecureStartupMessage
 			return conn, nil, newErrorf(code, "unsupported startup message: %T", m)
 		}
@@ -59,21 +63,22 @@ var FrontendAdmit = func(
 			return nil, nil
 		}
 		conn = tls.Server(conn, cfg)
+
+		// Now that SSL is established, read the encrypted startup message.
+		m, err = pgproto3.NewBackend(pgproto3.NewChunkReader(conn), conn).ReceiveStartupMessage()
+		if err != nil {
+			return conn, nil, newErrorf(codeClientReadFailed, "receiving post-TLS startup message: %v", err)
+		}
 	}
 
-	m, err := pgproto3.NewBackend(pgproto3.NewChunkReader(conn), conn).ReceiveStartupMessage()
-	if err != nil {
-		return conn, nil, newErrorf(codeClientReadFailed, "receiving post-TLS startup message: %v", err)
-	}
-	msg, ok := m.(*pgproto3.StartupMessage)
-	if !ok {
-		return conn, nil, newErrorf(codeUnexpectedStartupMessage, "unsupported post-TLS startup message: %T", m)
-	}
-
-	// Add the sniServerName (if used) as parameter
-	if sniServerName != "" {
-		msg.Parameters["sni-server"] = sniServerName
+	if startup, ok := m.(*pgproto3.StartupMessage); ok {
+		// Add the sniServerName (if used) as parameter
+		if sniServerName != "" {
+			startup.Parameters["sni-server"] = sniServerName
+		}
+		return conn, startup, nil
 	}
 
-	return conn, msg, nil
+	code := codeUnexpectedStartupMessage
+	return conn, nil, newErrorf(code, "unsupported post-TLS startup message: %T", m)
 }

--- a/pkg/ccl/sqlproxyccl/frontend_admitter_test.go
+++ b/pkg/ccl/sqlproxyccl/frontend_admitter_test.go
@@ -95,6 +95,35 @@ func TestFrontendAdmitWithClientSSLRequire(t *testing.T) {
 	require.NotNil(t, msg)
 }
 
+// TestFrontendAdmitRequireEncryption sends StartupRequest when SSlRequest is
+// expected.
+func TestFrontendAdmitRequireEncryption(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	cli, srv := net.Pipe()
+	require.NoError(t, srv.SetReadDeadline(timeutil.Now().Add(3e9)))
+	require.NoError(t, cli.SetReadDeadline(timeutil.Now().Add(3e9)))
+
+	go func() {
+		startup := pgproto3.StartupMessage{
+			ProtocolVersion: pgproto3.ProtocolVersionNumber,
+			Parameters:      map[string]string{"key": "val"},
+		}
+		_, err := cli.Write(startup.Encode([]byte{}))
+		require.NoError(t, err)
+	}()
+
+	tlsConfig, err := tlsConfig()
+	require.NoError(t, err)
+	frontendCon, msg, err := FrontendAdmit(srv, tlsConfig)
+	require.EqualError(t, err,
+		"codeUnexpectedInsecureStartupMessage: "+
+			"unsupported startup message: *pgproto3.StartupMessage")
+	require.NotNil(t, frontendCon)
+	require.Nil(t, msg)
+}
+
+// TestFrontendAdmitWithCancel sends CancelRequest.
 func TestFrontendAdmitWithCancel(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
@@ -109,14 +138,12 @@ func TestFrontendAdmitWithCancel(t *testing.T) {
 	}()
 
 	frontendCon, msg, err := FrontendAdmit(srv, nil)
-	require.EqualError(t, err,
-		"codeUnexpectedStartupMessage: "+
-			"unsupported post-TLS startup message: *pgproto3.CancelRequest",
-	)
+	require.NoError(t, err)
 	require.NotNil(t, frontendCon)
 	require.Nil(t, msg)
 }
 
+// TestFrontendAdmitWithSSLAndCancel sends SSLRequest followed by CancelRequest.
 func TestFrontendAdmitWithSSLAndCancel(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

--- a/pkg/ccl/sqlproxyccl/proxy_handler.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler.go
@@ -208,7 +208,8 @@ func (handler *proxyHandler) handle(ctx context.Context, incomingConn *proxyConn
 	}
 
 	// This currently only happens for CancelRequest type of startup messages
-	// that we don't support.
+	// that we don't support. Return nil to the server, which simply closes the
+	// connection.
 	if msg == nil {
 		return nil
 	}

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -31,7 +31,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -491,7 +490,6 @@ func TestDenylistUpdate(t *testing.T) {
 
 func TestDirectoryConnect(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 67729, "flaky test")
 
 	ctx := context.Background()
 	te := newTester()
@@ -606,27 +604,28 @@ func TestDirectoryConnect(t *testing.T) {
 	t.Run("drain connection", func(t *testing.T) {
 		url := fmt.Sprintf("postgres://root:admin@%s/?sslmode=disable&options=--cluster=tenant-cluster-28", addr)
 		te.TestConnect(ctx, t, url, func(conn *pgx.Conn) {
-			// Previous successful connection asynchronously closes - PGX cuts
-			// the connection and it can take time for the proxy to get the
-			// notification and react.
+			// The current connection count can take a bit of time to drop to 1,
+			// since the previous successful connection asynchronously closes.
+			// PGX cuts the connection on the client side, but it can take time
+			// for the proxy to get the notification and react.
 			require.Eventually(t, func() bool {
 				return proxy.metrics.CurConnCount.Value() == 1
 			}, 10*time.Second, 10*time.Millisecond)
 
-			// Connection should be terminated after draining.
+			// Connection should be forcefully terminated after the drain timeout,
+			// even though it's being continuously used.
 			require.Eventually(t, func() bool {
-				var n int
-				err = conn.QueryRow(ctx, "SELECT $1::int", 1).Scan(&n)
-				if err != nil {
-					return true
-				}
-				require.EqualValues(t, 1, n)
-
-				// Trigger drain of connections.
+				// Trigger drain of connections. Do this repeatedly inside the
+				// loop in order to avoid race conditions where the proxy is not
+				// yet hooked up to the directory server (and thus misses any
+				// one-time DRAIN notifications).
 				tds2.Drain()
-				return false
+
+				// Run query until it fails (because connection was closed).
+				return runTestQuery(ctx, conn) != nil
 			}, 30*time.Second, 5*drainTimeout)
 
+			// Ensure failure was due to forced drain disconnection.
 			require.Equal(t, int64(1), proxy.metrics.IdleDisconnectCount.Count())
 		})
 	})

